test/extensions/quic_listeners/quiche/envoy_quic_alarm_test.cc

```diff
@@ -0,0 +1,159 @@
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm1->IsSet());
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  alarm1->Cancel();
+  EXPECT_FALSE(alarm1->IsSet());
+  // Advance 10us, alarm1 shouldn't fire.
+  advanceUsAndLoop(10);
+  EXPECT_TRUE(unowned_delegate2->fired());
+  EXPECT_FALSE(unowned_delegate1->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, CreateAlarmAndReset) {
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  // Reset alarm1 to a different deadline.
+  alarm1->Cancel();
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(5));
+  // Advance 9us, alarm1 should have fired but alarm2 shouldn't.
+  advanceUsAndLoop(9);
+  EXPECT_TRUE(unowned_delegate1->fired());
+  EXPECT_FALSE(unowned_delegate2->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, CreateAlarmAndUpdate) {
+  auto unowned_delegate1 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm1(alarm_factory_.CreateAlarm(unowned_delegate1));
+  alarm1->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  auto unowned_delegate2 = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm2(alarm_factory_.CreateAlarm(unowned_delegate2));
+  alarm2->Set(clock_.Now() + QuicTime::Delta::FromMicroseconds(10));
+  EXPECT_TRUE(alarm2->IsSet());
+
+  // Update alarm1 to an earlier deadline.
+  alarm1->Update(clock_.Now() + QuicTime::Delta::FromMicroseconds(5),
+                 quic::QuicTime::Delta::Zero());
+  // Advance 9us, alarm1 should have fired but alarm2 shouldn't.
+  advanceUsAndLoop(9);
+  EXPECT_TRUE(unowned_delegate1->fired());
+  EXPECT_FALSE(unowned_delegate2->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, SetAlarmToPastTime) {
+  advanceUsAndLoop(100);
+  EXPECT_EQ(100, (clock_.Now() - quic::QuicTime::Zero()).ToMicroseconds());
+  auto unowned_delegate = new TestDelegate();
+  quic::QuicArenaScopedPtr<quic::QuicAlarm> alarm(alarm_factory_.CreateAlarm(unowned_delegate));
+  alarm->Set(clock_.Now() - QuicTime::Delta::FromMicroseconds(10));
+  base_scheduler_.run(Dispatcher::RunType::NonBlock);
+  EXPECT_TRUE(unowned_delegate->fired());
+}
+
+TEST_F(EnvoyQuicAlarmTest, FireAlarmWithPastDeadline) {
```

Comment:
This test is identical to SetAlarmToPastTime. Maybe change it to UpdateAlarmToPastTime and test that Update() to a past time will fire the alarm immediately?